tools/clang/lib/SPIRV/SpirvEmitter.cpp

```diff
@@ -13246,6 +13246,31 @@ void SpirvEmitter::processInlineSpirvAttributes(const FunctionDecl *decl) {
   }
 }
 
+bool SpirvEmitter::processNumThreadsAttr(const FunctionDecl *decl) {
+  auto *numThreadsAttr = decl->getAttr<HLSLNumThreadsAttr>();
+  if (!numThreadsAttr)
+    return false;
+
+  auto f = [](ASTContext &Ctx, Expr *E) {
+    if (E) {
+      llvm::APSInt apsInt;
+      APValue apValue;
+      if (E->isIntegerConstantExpr(apsInt, Ctx))
+        return (uint32_t)apsInt.getSExtValue();
+      if (E->isVulkanSpecConstantExpr(Ctx, &apValue) && apValue.isInt())
+        return (uint32_t)apValue.getInt().getSExtValue();
+    }
+    return 1U;
+  };
+
+  spvBuilder.addExecutionMode(entryFunction, spv::ExecutionMode::LocalSize,
+                              {f(astContext, numThreadsAttr->getX()),
+                               f(astContext, numThreadsAttr->getY()),
+                               f(astContext, numThreadsAttr->getZ())},
+                              decl->getLocation());
+  return true;
+}
```

Comment:
If you have a spec constant, you need to use the `LocalSizeId` execution mode, with the id of the spec constant right?

We could implement this by always using `LocalSizeId`, and make the parameters the id of the expression in the attribute.